File: src/collector-client.mjs
import { execFile } from "node:child_process";
import { promisify } from "node:util";

export function validateCollectorUrl(value, tunnelId) {
  const url = new URL(value);
  if (url.username || url.password || url.search || url.hash || url.pathname !== "/") throw new Error("Use the ingestion origin without a path or credentials");
  const local = url.protocol === "http:" && ["127.0.0.1", "localhost", "[::1]"].includes(url.hostname);
  const remote = url.protocol === "https:" && /^[a-z0-9-]+\.[a-z0-9]+\.devtunnels\.ms$/i.test(url.hostname);
  if (!local && !remote) throw new Error("Use an HTTPS Dev Tunnel or HTTP loopback URL");
  if (remote && (!/^[a-z0-9][a-z0-9-]*\.[a-z0-9]+$/i.test(tunnelId || "") || !url.hostname.endsWith("." + tunnelId.split(".").pop() + ".devtunnels.ms"))) throw new Error("The ingestion tunnel ID and cluster are required");
  return url;
}

export function createCollectorClient(settings, { request = fetch, execute = promisify(execFile), now = Date.now } = {}) {
  const origin = validateCollectorUrl(settings.url, settings.tunnelId);
  let tunnelToken = null; let tokenExpires = 0;
  async function connectToken() {
    if (tunnelToken && tokenExpires > now()) return tunnelToken;
    try {
      const result = await execute(process.platform === "win32" ? "devtunnel.exe" : "devtunnel", ["token", settings.tunnelId, "--scopes", "connect", "--json"], { timeout: 60000, windowsHide: true });
      const parsed = JSON.parse(result.stdout);
      tunnelToken = typeof parsed.token === "string" ? parsed.token : parsed.accessToken || parsed.token?.token || parsed.token?.accessToken;
      if (typeof tunnelToken !== "string") throw new Error();
      tokenExpires = now() + 60 * 60 * 1000;
      return tunnelToken;
    } catch { throw new Error("Dev Tunnel sign-in failed. Run devtunnel user login --entra on this device."); }
  }
  return {
    async send(route, method, body, token) {
      const headers = { "content-type": "application/json", Accept: "application/json" };
      if (origin.protocol === "https:") headers["X-Tunnel-Authorization"] = "tunnel " + await connectToken();
      if (token) headers.Authorization = "Bearer " + token;
      const destination = new URL(route, origin);
      if (destination.origin !== origin.origin) throw new Error("Collector routes must stay on the configured ingestion origin");
      const response = await request(destination, { method, headers, body: JSON.stringify(body), redirect: "error", signal: AbortSignal.timeout(30000) });
      if ([401, 403].includes(response.status)) { tunnelToken = null; tokenExpires = 0; }
      if (!response.ok) throw Object.assign(new Error("Collector request failed (" + response.status + "). Check pairing, sign-in, and ingestion URL."), { status: response.status });
      return response.json();
    }
  };
}

File: src/Dashboard.Core/CollectorClient.cs
using System.Diagnostics;
using System.Net.Http.Json;
using System.Text.Json;
using Dashboard.Contracts;

namespace Dashboard.Core;

public sealed class CollectorClient : IDisposable
{
    private readonly CollectorConfig config;
    private readonly HttpClient http;
    private string? tunnelToken;
    private DateTimeOffset tokenExpires;
    public CollectorClient(CollectorConfig config)
    {
        this.config = config;
        var url = new Uri(config.Url);
        if ((!url.IsLoopback && url.Scheme != "https") || url.Scheme is not ("http" or "https") || url.UserInfo != "" || url.AbsolutePath != "/" || url.Query != "" || url.Fragment != "")
            throw new InvalidDataException("Collector URL must be an HTTPS origin or HTTP loopback origin");
        http = new(new HttpClientHandler { AllowAutoRedirect = false }) { BaseAddress = url, Timeout = TimeSpan.FromSeconds(30) };
    }
    public async Task<T> Send<T>(string route, HttpMethod method, object? body, bool authenticate, CancellationToken token)
    {
        if (!route.StartsWith('/') || route.StartsWith("//", StringComparison.Ordinal)) throw new InvalidOperationException("Relative collector route required");
        using var request = new HttpRequestMessage(method, route);
        if (body is not null) request.Content = JsonContent.Create(body, options: Protocol.Json);
        request.Headers.Accept.ParseAdd("application/json");
        if (authenticate) request.Headers.Authorization = new("Bearer", config.Token);
        if (!string.IsNullOrEmpty(config.TunnelId))
        {
            if (tunnelToken is null || DateTimeOffset.UtcNow >= tokenExpires) await RefreshTunnelToken(token);
            request.Headers.Add("X-Tunnel-Authorization", "tunnel " + tunnelToken);
        }
        using var response = await http.SendAsync(request, token);
        if (!response.IsSuccessStatusCode)
        {
            if ((int)response.StatusCode is 401 or 403) tunnelToken = null;
            throw new DomainException((int)response.StatusCode, $"Collector HTTP {(int)response.StatusCode}; check sign-in and enrollment");
        }
        return await response.Content.ReadFromJsonAsync<T>(Protocol.Json, token) ?? throw new InvalidDataException("Empty API response");
    }
    private async Task RefreshTunnelToken(CancellationToken token)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(token); timeout.CancelAfter(TimeSpan.FromSeconds(60));
        var start = new ProcessStartInfo(OperatingSystem.IsWindows() ? "devtunnel.exe" : "devtunnel") { RedirectStandardOutput = true, RedirectStandardError = true, CreateNoWindow = true, UseShellExecute = false };
        foreach (var value in new[] { "token", config.TunnelId!, "--scopes", "connect", "--json" }) start.ArgumentList.Add(value);
        using var process = Process.Start(start) ?? throw new InvalidOperationException("Dev Tunnels CLI unavailable");
        var output = process.StandardOutput.ReadToEndAsync(timeout.Token); var error = process.StandardError.ReadToEndAsync(timeout.Token);
        try
        {
            await process.WaitForExitAsync(timeout.Token); await error;
            if (process.ExitCode != 0) throw new InvalidOperationException("Run devtunnel user login on this device");
            using var doc = JsonDocument.Parse(await output); var root = doc.RootElement;
            tunnelToken = root.TryGetProperty("token", out var t) && t.ValueKind == JsonValueKind.String ? t.GetString() : root.GetProperty("accessToken").GetString();
            tokenExpires = DateTimeOffset.UtcNow.AddHours(1);
        }
        finally { if (!process.HasExited) process.Kill(true); }
    }
    public void Dispose() => http.Dispose();
}
